java/core/lance-jni/src/error.rs

```diff
@@ -15,149 +15,171 @@
         }
     }
 
-    /// Throw as an concrete Java Exception
-    pub fn throw_as(&self, env: &mut jni::JNIEnv, exception: JavaException) {
-        env.throw_new(exception.as_str(), self.to_string())
+    pub fn io_error(message: String) -> Self {
+        Self::new(message, JavaExceptionClass::IOException)
+    }
+
+    pub fn input_error(message: String) -> Self {
+        Self::new(message, JavaExceptionClass::IllegalArgumentException)
+    }
+
+    pub fn unsupported_error(message: String) -> Self {
+        Self::new(message, JavaExceptionClass::UnsupportedOperationException)
+    }
+
+    pub fn throw(&self, env: &mut JNIEnv) {
+        env.throw_new(self.java_class.as_str(), &self.message)
             .expect("Error when throwing Java exception");
     }
 }
 
-pub type Result<T> = std::result::Result<T, Error>;
+impl std::fmt::Display for JavaError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}: {}", self.java_class.as_str(), self.message)
+    }
+}
+
+impl std::error::Error for JavaError {}
+
+/// Trait for converting errors to Java exceptions.
+pub trait JavaErrorConversion<T> {
+    /// Convert to `JavaError` as I/O exception.
+    fn io_error(self) -> JavaResult<T>;
 
-trait ToSnafuLocation {
-    fn to_snafu_location(&'static self) -> snafu::Location;
+    /// Convert to `JavaError` as runtime exception.
+    fn runtime_error(self) -> JavaResult<T>;
+
+    /// Convert to `JavaError` as value (input) exception.
+    fn input_error(self) -> JavaResult<T>;
+
+    /// Convert to `JavaError` as unsupported operation exception.
+    fn unsupported_error(self) -> JavaResult<T>;
 }
 
-impl ToSnafuLocation for std::panic::Location<'static> {
-    fn to_snafu_location(&'static self) -> snafu::Location {
-        snafu::Location::new(self.file(), self.line(), self.column())
+impl<T, E: std::fmt::Display> JavaErrorConversion<T> for std::result::Result<T, E> {
+    fn io_error(self) -> JavaResult<T> {
+        self.map_err(|err| JavaError::io_error(err.to_string()))
+    }
+
+    fn runtime_error(self) -> JavaResult<T> {
+        self.map_err(|err| JavaError::runtime_error(err.to_string()))
+    }
+
+    fn input_error(self) -> JavaResult<T> {
+        self.map_err(|err| JavaError::input_error(err.to_string()))
+    }
+
+    fn unsupported_error(self) -> JavaResult<T> {
+        self.map_err(|err| JavaError::unsupported_error(err.to_string()))
     }
 }
 
-impl From<JniError> for Error {
-    #[track_caller]
-    fn from(source: JniError) -> Self {
-        Self::Jni {
-            message: source.to_string(),
-            location: std::panic::Location::caller().to_snafu_location(),
+/// JavaErrorExt trait that converts specific error types to Java exceptions
+pub trait JavaErrorExt<T> {
+    /// Convert to a Java error based on the specific error type
+    fn infer_error(self) -> JavaResult<T>;
+}
+
+impl<T> JavaErrorExt<T> for std::result::Result<T, LanceError> {
```

Comment:
This is first time i saw the conversion on Result instead of Error.  Why do we do this?

Reply:
Mimic the Python side to reduce the error conversion headache, we do not need to list out all the errors and create the message for each error again, makes the error conversion much easier (keep the original info while can convert to target java exception)